### ISSUE TITLE
Support constructor annotations

### ISSUE THREAD

**Issue by rgabbard-bbn** [immutables/immutables#317]
It would be nice to provide some means of specifying annotations to be copied to the generated constructor. This would be useful for allowing generated objects to have their parameters injected by Guice by placing `@Inject` on the generated constructor.

**Comment by elucash** [comment]
We have similar option for propagating annotations to accessors, I'll try to adjust it to propagate to constuctors if annotation matches `Target.PARAMETER` and mentioned in `passAnnotations` style.

**Comment by elucash** [comment]
released in 2.2-beta-2
use `Style.passAnnotations` to specify all annotations you might want to pass.

```java
@Value.Style(passAnnotations = {Inject.class}, of = "new") // assuming public constructors
package top.package.whatever;
```

```java
@Value.Immutable
interface Service {
  @Value.Parameter @Inject Dependency dependency();
  @Value.Parameter @Inject Collaborator collaborator();
}
...
// Generated constructor will have annotations which
// have target `ElementType.PARAMETER` added
Service(@Inject Dependency dependency, @Inject Collaborator collaborator) {
  ....
}
```

**Comment by rgabbard-bbn** [comment]
Thanks! This makes a lot of progress towards what I needed, but one thing is still lacking (I should have provided an example).  Typically when injecting a class with Guice it will look like this:

```java
final class MyClass {
    @Inject
    MyClass(@SomeAnnotation1 DepClass1 dep1, @SomeAnnotation2 DepClass2 dep2) {
         ...
    }
}
```

This fix is great for applying the parameter annotations; is there a way to apply the `@Inject` annotation to the generated constructor itself?

**Comment by elucash** [comment]
Haven't used DI for quite a while, even forgot such a details. While passing arbitrary annotations to parameters should work, the constructor itself should be somehow annotated. Do you have an example annotation API on your mind, a one which would not look too ugly?

**Comment by rgabbard-bbn** [comment]
The API for this is tricky.  Something like `@Value.Style(constructorAnnotations={Inject.class})` would work for "parameterless" constructor annotations, which is sufficient for what Guice needs.

I don't think you can specify annotations with parameters in a similar way, though.  We would need some sort of carrier like:

```java
class MyClass {
    @Value.ConstructorProxy
    @MyConstructorAnnotation(param="foo")
    protected abstract void constructorProxy();
}
```

which is a little uglier.

**Comment by elucash** [comment]
Good idea! How about having actual no-arg constructor as annotation holder?

```java
class MyClass {
   @MyConstructorAnnotation(param="foo")
   MyClass() {}
}
```

**Comment by rgabbard-bbn** [comment]
Sounds good!

**Comment by elucash** [comment]
2.2-beta-3 is here to try

```java
@Value.Immutable
abstract class Serv {
   @Inject Serv() {}
   @Value.Parameter A a();
   @Value.Parameter B b();
} 
```

**Comment by rgabbard-bbn** [comment]
@elucash : Thanks! I tried applying both the parameter and constructor annotations to a class in my project and used Maven's `dependencyManagement` to use `2.2-beta-3` for all my Immutables dependencies.  What I am seeing is that the annotations on the parameters are copied onto the accessors of the generated class but not onto the constructor parameters, and the annotation on the constructor is not propagated.    I will try to see if I can make a minimal example disentangled from our code to replicate the problem sometime soon.

Trying it out also raised a concern about the API for parameter annotations. Because styles are not merged, you can't use `@Value.Style(passAnnotations = {Inject.class})` together with a custom style, which is quite inconvenient (in Guice at least the annotations applied to constructor arguments are typically different for each class).

One option would be to allow putting annotated parameters on the template class constructor  and then propagating those annotations (assuming that information is available to the annotation processor).  That provides a nice interface for the user, but having the generated class pass dummy values to a parent constructor is distasteful.  Another option would be to go back to the route of having some dummy method stand in for the constructor, which is less elegant for the user but less weird in the implementation.

**Comment by elucash** [comment]
found a problem in code which handles meta-annotations. Adsense of style merging is an inconvenience, for sure, but the problem with the need to specify `passAnnotations` is usually solved by listing meta-annotations. In case of Guice, `javax.annotation.Qualifier` meta-annotation can be specified and it will automatically include any qualifier annotation that is annotated with `javax.annotation.Qualifier`.
So with newly released `2.2-beta-5`, the following should work:

```java
// package-info.java
@org.immutables.value.Value.Style(passAnnotations = {javax.inject.Qualifier.class}, of = "new")
package guicesample;
```

```java
// Secured.java
package guicesample;

import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Retention;
import javax.inject.Qualifier;

@Qualifier
@Retention(RetentionPolicy.RUNTIME)
public @interface Secured {}
```

```java
// Serv.java
package guicesample;

import org.immutables.value.Value;
import javax.inject.Inject;

@Value.Immutable
public abstract class Serv {
	@Inject Serv() {}

	@Value.Parameter
	public abstract @Secured Vault vault();
}
```

And this will result in the code generated

```java
// ImmutableServ.java
...
@SuppressWarnings("all")
@ParametersAreNonnullByDefault
@Generated({"Immutables.generator", "Serv"})
@Immutable
public final class ImmutableServ extends Serv {
  private final Vault vault;

  /**
   * Construct a new immutable {@code Serv} instance.
   * @param vault The value for the {@code vault} attribute
   */
  @Inject
  public ImmutableServ(@Secured Vault vault) {
    this.vault = Preconditions.checkNotNull(vault, "vault");
  }

  private ImmutableServ(ImmutableServ original, Vault vault) {
    this.vault = vault;
  }
  ....
```

**Comment by elucash** [comment]
released in 2.2 Thank you!
